### DISCUSSION TITLE
我需要为每个页面提供不同的 script 并使用

### DISCUSSION THREAD

**Discussion by Aomd** [umijs/umi#11188] [Q&A]
# 配置
``` ts
// "umi": "^4.0.69"

export default defineConfig({
  title: false,
  hash: true,
  fastRefresh: true,
  exportStatic: {},
  ssr: {},
});

```
# 代码
```ts
import { useEffect } from "react";
import { Helmet } from "umi";

export default function app() {

  useEffect(() => {
    console.log('uuid', window.uuid)
  }, [])

  return (
    <>
      <Helmet>
        <title> test-umi </title>
        <script>console.log('start')</script>
        <script src="https://lf6-cdn-tos.bytecdntp.com/cdn/expire-1-M/uuid/8.3.2/uuid.min.js" type="application/javascript" ></script>
      </Helmet>
    </>
  );
}
```
# 输出结果
``` shell
uuid undefined
start
```

# 期望结果
``` shell
start
uuid {NIL: '00000000-0000-0000-0000-000000000000', parse: ƒ, stringify: ƒ, v1: ƒ, v3: ƒ, …}
```

**Comment by fz6m** [comment]
判断一下路径，条件渲染 Helmet 。

**Reply by Aomd** [reply]
开启了 ssr Helmet 东西会添加到 head里面 按照执行顺序 useEffect hook 里面应该会打印出 uuid 这个类的值

**Reply by fz6m** [reply]
如有问题，需要给一个最小复现。

**Reply by Aomd** [reply]
你好 复现地址为 https://gitee.com/Aomd/umi-example

赶上“毕业季“ 加上目前还是以 3.x 为主 所以延后了一个半月 非常抱歉

**Reply by fz6m** [reply]
因为 umi 4 是 react 18 ，默认用了 react 18 的流式渲染，在开发时是 ssr ，所以没等 helmet 运行就已经边渲染边传给用户了，自然访问到 html 一开始没有 script 头，把构建出来的 ssr 产物运行起来也是和开发时一样的效果。如果先运行一遍组件收集 helmet 就是全量了，失去了 stream 的意义。

但是构建的时候开启了 `exportStatic` ，这是类似于 ssg 的功能，会做 react render to string ，这个时候 helmet 被提取出来了，产物的 html 里就有 script 头了。

推荐以 npm 包的形式导入，而不是 umd 的形式引入，如果不想浪费自己的流量需要用 umd ，就配到全局的 [scripts](https://umijs.org/docs/api/config#scripts) 里，虽然没法按页面精细化，但不差这点。

**Reply by fz6m** [reply]
在最新版本中，有一个 `useServerInsertedHTML` 可以给 html 插入标签，我试了下是可以满足你的需求的：

```ts
import { useServerInsertedHTML } from "umi";

  useServerInsertedHTML(() => {
    return <script src="https://lf6-cdn-tos.bytecdntp.com/cdn/expire-1-M/uuid/8.3.2/uuid.min.js" type="application/javascript" />
  })
```

**Reply by Aomd** [reply]
首先感谢大佬百忙之中抽出时间来解答、非常感谢，其次回到问题上来
1、我的应用场景是多页面（umi 约定路由 pages 下 xx.tsx 都是一个页面） 非单页面应用程序（Single Page Application，SPA）第三方的 script 会很多 并且每个页面都会有各自的 script 所以 umd 很有必要
2、useServerInsertedHTML 我尝试了下 会将 script 添加到 html 末尾 dev 和 build 都一样 在浏览器缓存下 确实可以正确打印出来 但是
disable cache 之后 因为浏览器加载机制 还是会导致 uuid 为 undefined

so，第三方的 script 最好的解决方案还是 按照浏览器加载机制添加到每个页面的 head 中
ps：不是不想浪费自己流量 单纯是为了示例的便捷性（年轻的时候已经被第三方免费 cdn 坑怕了 哪怕是大厂）
